### PR TITLE
[BEAM-77] Move InProcessRunner to its own module

### PR DESCRIPTION
**Pull request by kennknowles**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [x] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [x] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---

I believe since the `DirectPipelineRunner` is still in place, this safe to do immediately. Then we can begin to run all tests on this runner in the same way we do for all runners, gaining coverage of both. Finally, we can simply delete `DirectPipelineRunner` which will expose any tests that require the `NeedsRunner` category.

### CODE REVIEW

**Comment by kennknowles** [comment]
R: @tgroh 
R: @davorbonaci 

Perhaps just high level comments on the approach. Specifically, do you agree on this order of things:

1. Do this move now.
2. Leverage `TestPipeline` to immediately increase test coverage to include this runner.
3. Annotate tests appropriately with `NeedsRunner`
4. Delete `DirectPipelineRunner` from the SDK.

Expect some hiccups along the way. I'm sure that we will have to learn during set up of item 2, and that item 4 will feed back into item 3 since we will miss some.

**Comment by davorbonaci** [comment]
Yes, I think I agree with a few minor notes, such as separating into smaller PRs.

We may be able to avoid conflating addition of new code (e.g., `NeedsRunner`) with the move itself. Could be easier to review and reason about it that way. (This is particularly because I'm not sure all tests will actually work on the new runner right now -- they may soon.)

We should coordinate with pending PRs from @tgroh. I'm fine whichever order you decide to proceed in.

Note that the official name of the runner will be `DirectRunner`, without `Pipeline` or `InProcess`. Naming-wise, I'd probably go with:
* Directory: `/runners/direct-java`
* Artifact: `direct-runner`
* Java package: `org.apache.beam.runners.direct.*`

**Comment by kennknowles** [comment]
Done:

 - Removed the not-yet-used `NeedsRunner` commit from this PR.
 - Since that required a rebase, I also moved the commit on top of `apache/master`.
 - Named directory, artifact, and package as you describe.

This commit is now strictly a move with a `pom.xml`. I don't think it is helpful to split this into two, since the dependency analysis requires the code to be in place to guide the `pom.xml`.

From here out, the changes will need tighter coordination to bring up the new runner with its appropriate name and running all the integration tests appropriately.

**Comment by kennknowles** [comment]
Rebased, fixed conflicts, Jenkins is happy. Please take a look!

**Comment by kennknowles** [comment]
I believe Travis-CI failures are memory/performance issues in `InProcessPipelineRunner` not related to this PR, since there are no code changes here.

**Comment by kennknowles** [comment]
Rebased to kick the tests.

File: runners/direct-java/pom.xml [hunk 1]
```diff
@@ -0,0 +1,430 @@
...
+  </parent>
+
+  <artifactId>direct-runner</artifactId>
+
+  <name>Apache Beam :: Runners :: Direct</name>
+
+  <packaging>jar</packaging>
+
+  <properties>
+    <beam.version>0.1.0-incubating-SNAPSHOT</beam.version>
```
Comment (davorbonaci) [inline runners/direct-java/pom.xml]: remove. use ${project.version} instead.

File: runners/direct-java/pom.xml [hunk 2]
```diff
@@ -0,0 +1,430 @@
...
+    <timestamp>${maven.build.timestamp}</timestamp>
+    <maven.build.timestamp.format>yyyy-MM-dd HH:mm</maven.build.timestamp.format>
```
Comment (davorbonaci) [inline runners/direct-java/pom.xml]: can we remove these? Aren't they inherited from the parent?

File: runners/direct-java/pom.xml [hunk 3]
```diff
@@ -0,0 +1,430 @@
...
+        <executions>
+          <!-- In the first phase, we pick dependencies and relocate them. -->
+          <execution>
+            <id>bundle-and-repackage</id>
+            <phase>package</phase>
+            <goals>
+              <goal>shade</goal>
+            </goals>
+            <configuration>
+              <shadeTestJar>true</shadeTestJar>
```
Comment (davorbonaci) [inline runners/direct-java/pom.xml]: remove much of these?

File: runners/direct-java/pom.xml [hunk 4]
```diff
@@ -0,0 +1,430 @@
...
+  </build>
+
+  <dependencies>
+    <dependency>
+      <groupId>org.apache.beam</groupId>
+      <artifactId>java-sdk-all</artifactId>
+      <version>${beam.version}</version>
+    </dependency>
+
+    <dependency>
```
Comment (davorbonaci) [inline runners/direct-java/pom.xml]: why do we need all these?

File: runners/direct-java/pom.xml [hunk 5]
```diff
@@ -0,0 +1,430 @@
...
+      <groupId>org.apache.beam</groupId>
+      <artifactId>java-sdk-all</artifactId>
+      <version>${beam.version}</version>
+      <type>test-jar</type>
+      <scope>test</scope>
+    </dependency>
+
+    <dependency>
```
Comment (davorbonaci) [inline runners/direct-java/pom.xml]: probably not needed

File: runners/direct-java/src/main/java/org/apache/beam/runners/direct/ParDoMultiEvaluatorFactory.java [hunk 6]
```diff
@@ -15,8 +15,8 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package org.apache.beam.sdk.runners.inprocess;
-import org.apache.beam.sdk.runners.inprocess.InProcessPipelineRunner.CommittedBundle;
+package org.apache.beam.runners.direct;
+import org.apache.beam.runners.direct.InProcessPipelineRunner.CommittedBundle;
```
Comment (davorbonaci) [inline runners/direct-java/src/main/java/org/apache/beam/runners/direct/ParDoMultiEvaluatorFactory.java]: empty line?

File: runners/direct-java/src/test/java/org/apache/beam/runners/direct/ParDoMultiEvaluatorFactoryTest.java [hunk 7]
```diff
@@ -119,7 +119,7 @@ public void processElement(ProcessContext c) {
     CounterSet counters = new CounterSet();
     when(evaluationContext.createCounterSet()).thenReturn(counters);
 
-    org.apache.beam.sdk.runners.inprocess.TransformEvaluator<String> evaluator =
+    org.apache.beam.runners.direct.TransformEvaluator<String> evaluator =
```
Comment (davorbonaci) [inline runners/direct-java/src/test/java/org/apache/beam/runners/direct/ParDoMultiEvaluatorFactoryTest.java]: why this is not an import? (ok to ignore)

File: runners/pom.xml [hunk 8]
```diff
@@ -37,6 +37,7 @@
   <modules>
     <module>flink</module>
```
Comment (davorbonaci) [inline runners/pom.xml]: move up; direct runner should be first

**Comment by davorbonaci** [comment]
Looks mostly good. Advance LGTM.

File: runners/direct-java/pom.xml (hunk 1 above)
Comment (kennknowles) [inline runners/direct-java/pom.xml]: Done.

File: runners/direct-java/pom.xml (hunk 2 above)
Comment (kennknowles) [inline runners/direct-java/pom.xml]: Done.

File: runners/direct-java/pom.xml (hunk 3 above)
Comment (kennknowles) [inline runners/direct-java/pom.xml]: Done.

File: runners/direct-java/pom.xml (hunk 4 above)
Comment (kennknowles) [inline runners/direct-java/pom.xml]: Let's trim `google-http-client` later and see why dependency analysis was OK with it. It is `compile` scope so it is definitely handled by the analysis.

File: runners/direct-java/pom.xml (hunk 5 above)
Comment (kennknowles) [inline runners/direct-java/pom.xml]: Done. It seems the same flag (`ignoreNonCompile`) that makes it not fail on `provided` and `runtime` dependencies - basically mandatory to set this - encompasses `test` as well. We can keep digging but maybe not for this PR.

File: runners/direct-java/src/main/java/org/apache/beam/runners/direct/ParDoMultiEvaluatorFactory.java (hunk 6 above)
Comment (kennknowles) [inline runners/direct-java/src/main/java/org/apache/beam/runners/direct/ParDoMultiEvaluatorFactory.java]: Done.

File: runners/direct-java/src/test/java/org/apache/beam/runners/direct/ParDoMultiEvaluatorFactoryTest.java (hunk 7 above)
Comment (kennknowles) [inline runners/direct-java/src/test/java/org/apache/beam/runners/direct/ParDoMultiEvaluatorFactoryTest.java]: No idea. No need to import even; same package. Done.

File: runners/pom.xml (hunk 8 above)
Comment (kennknowles) [inline runners/pom.xml]: Done.

**Comment by kennknowles** [comment]
Based on advance LGTM I will merge if/when tests are green. Cutting deps and maven properties may have busted something so we'll see.

**Comment by kennknowles** [comment]
Failure were only with Java 7 JDKs. The PermGen no longer exists in Java 8, so it isn't getting blown. I've adjusted the JVM parameters to settings that will hopefully alleviate the problem. I don't know why this problem does not occur when the runner is part of the SDK.

**Comment by davorbonaci** [comment]
@tgroh, is this something you've seen before?

**Pull request merged**